x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts

```diff
@@ -0,0 +1,744 @@
+      definition: rootStreamDefinition,
+    });
+
+    return { acknowledged: true, result: 'created' };
+  }
+
+  /**
+   * Disabling streams means deleting the logs root stream
+   * AND its descendants, including any Elasticsearch objects,
+   * such as data streams. That means it deletes all data
+   * belonging to wired streams.
+   *
+   * It does NOT delete ingest streams.
+   */
+  async disableStreams(): Promise<DisableStreamsResponse> {
+    const isEnabled = await this.isStreamsEnabled();
+    if (!isEnabled) {
+      return { acknowledged: true, result: 'noop' };
+    }
+
+    const definition = await this.getStream(LOGS_ROOT_STREAM_NAME);
+
+    await this.deleteStreamFromDefinition(definition);
+
+    return { acknowledged: true, result: 'deleted' };
+  }
+
+  /**
+   * Resyncing streams means re-installing all Elasticsearch
+   * objects (index and component templates, pipelines, and
+   * assets), using the stream definitions as the source of
+   * truth.
+   *
+   * Streams are re-synced in a specific order:
+   * the leaf nodes are synced first, then its parents, etc.
+   * Thiis prevents us from routing to data streams that do
+   * not exist yet.
+   */
+  async resyncStreams(): Promise<ResyncStreamsResponse> {
+    const streams = await this.getManagedStreams();
+
+    const streamsWithDepth = streams.map((stream) => ({
+      stream,
+      depth: stream.name.match(/\./g)?.length ?? 0,
+    }));
+
+    const streamsInOrder = orderBy(streamsWithDepth, 'depth', 'desc');
+
+    for (const { stream } of streamsInOrder) {
+      await this.syncStreamObjects({
+        definition: stream,
+      });
+    }
+    return { acknowledged: true, result: 'updated' };
+  }
+
+  private async syncStreamObjects({ definition }: { definition: StreamDefinition }) {
+    const { assetClient, logger, scopedClusterClient } = this.dependencies;
+
+    if (isWiredStream(definition)) {
+      await syncWiredStreamDefinitionObjects({
+        definition,
+        logger,
+        scopedClusterClient,
+      });
+    } else if (isIngestStream(definition)) {
+      await syncIngestStreamDefinitionObjects({
+        definition,
+        scopedClusterClient,
+        logger,
+        dataStream: await this.getDataStream(definition.name),
+      });
+    }
+
+    await assetClient.syncAssetList({
+      entityId: definition.name,
+      entityType: 'stream',
+      assetType: 'dashboard',
+      assetIds: definition.dashboards ?? [],
+    });
+  }
+
+  /**
+   * Creates or updates a stream. The routing of the parent is
+   * also updated (including syncing to Elasticsearch).
+   */
+  async upsertStream({
+    definition,
+  }: {
+    definition: StreamDefinition;
+  }): Promise<UpsertStreamResponse> {
+    const { result, parentDefinition } = await this.validateAndUpsertStream({
+      definition,
+    });
+
+    if (parentDefinition) {
+      const isRoutingToChild = parentDefinition.stream.ingest.routing.find((item) => item.name);
+
```

Comment:
The `find` doesn't compare the `item.name`, should be `item.name === definition.name`. I'll push a fix

Reply:
gee, thanks - I'll add a test for that one as well to https://github.com/elastic/streams-program/issues/26